lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs

```diff
@@ -1323,6 +1326,48 @@ prop_balanceTransactionExistingTotalCollateral
   where
     pp = mockPParamsForBalancing
 
+-- If 'balanceTx' is able to balance a transaction, then repeating the attempt
+-- with all potential inputs removed (from the partial transaction and the
+-- UTxO set) should always fail with 'ErrBalanceTxUnableToCreateInput'.
+--
+-- Note that we /cannot/ expect 'balanceTx' to fail with
+-- 'ErrBalanceTxUnableToCreateInput' in all situations where there are no
+-- potential inputs available, since:
+--
+-- 1. only at most one failure condition is ever reported by 'balanceTx';
+-- 2. there can be multiple competing failure conditions;
+-- 3. the order in which failure conditions are checked is unspecified.
+--
+prop_balanceTransactionUnableToCreateInput
+    -- TODO: Test with all recent eras [ADP-2997]
+    :: forall era. era ~ Write.BabbageEra
+    => Wallet
+    -> PartialTx era
+    -> StdGenSeed
+    -> Property
+prop_balanceTransactionUnableToCreateInput w ptx seed =
+    case r0 of
+        Right _balancedTx -> r1 === Left ErrBalanceTxUnableToCreateInput
+        Left _someFailure -> property True
+    & cover 1
+        (isRight r0)
+        "isRight r0"
+    & checkCoverage
+  where
+    r0 = balanceTx w pp tt seed ptx
+    r1 = balanceTx
+        (eraseWalletUTxOSet w) pp tt seed (erasePartialTxInputList ptx)
```

Comment:
Style topic: why not inline `r0` and `r1`, potentially factoring out `balanceTx' = balanceTx pp tt seed` modulo argument order? Maybe we can briefly discuss tomorrow

Reply:
This was resolved by f2cde3177d514381add96a5a4672e18d64182edd.